File: solutions/cpp/palindrome-partitioning.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <random>
#include <string>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::string;
using std::uniform_int_distribution;
using std::vector;

bool IsPalindrome(const string& s);
void PalindromePartitioningHelper(const string& s, size_t begin,
                                  vector<string>* partition,
                                  vector<vector<string>>* result);

// @include
vector<vector<string>> PalindromePartitioning(const string& s) {
  vector<vector<string>> result;
  vector<string> partition;
  PalindromePartitioningHelper(s, 0, &partition, &result);
  return result;
}

void PalindromePartitioningHelper(const string& s, size_t begin,
                                  vector<string>* partition,
                                  vector<vector<string>>* result) {
  if (begin == s.size()) {
    result->emplace_back(*partition);
    return;
  }

  for (size_t i = begin + 1; i <= s.size(); ++i) {
    string prefix = s.substr(begin, i - begin);
    if (IsPalindrome(prefix)) {
      partition->emplace_back(prefix);
      PalindromePartitioningHelper(s, i, partition, result);
      partition->pop_back();
    }
  }
}

bool IsPalindrome(const string& s) {
  for (int i = 0, j = s.size() - 1; i < j; ++i, --j) {
    if (s[i] != s[j]) {
      return false;
    }
  }
  return true;
}
// @exclude

void CheckAns(const vector<vector<string>> &vecs, const string& input) {
  for (const auto& vec : vecs) {
    string temp;
    for (const auto& s : vec) {
      assert(IsPalindrome(s));
      temp += s;
    }
    assert(!temp.compare(input));
  }
}

string RandString(size_t len) {
  default_random_engine gen((random_device())());
  string ret;
  uniform_int_distribution<char> dis('a', 'z');
  while (len) {
    --len;
    ret += dis(gen);
  }
  return ret;
}

int main(int argc, char** argv) {
  if (argc == 2) {
    string s = argv[1];
  } else {
    default_random_engine gen((random_device())());
    for (int times = 0; times < 1000; ++times) {
      uniform_int_distribution<size_t> dis_len(0, 10);
      string s = RandString(dis_len(gen));
      auto result = PalindromePartitioning(s);
      CheckAns(result, s);
      cout << "string s = " << s << endl;
      for (const auto& vec : result) {
        for (const auto& t : vec) {
          cout << t << " ";
        }
        cout << endl;
      }
    }
  }
  return 0;
}

File: solutions/python/palindrome_partitioning.py
import sys
import random
import string


# @include
def palindrome_partitioning(input):
    def directed_palindrome_partitioning(offset, partial_partition):
        if offset == len(input):
            result.append(list(partial_partition))
            return

        for i in range(offset + 1, len(input) + 1):
            prefix = input[offset:i]
            if prefix == prefix[::-1]:
                directed_palindrome_partitioning(i,
                                                 partial_partition + [prefix])

    result = []
    directed_palindrome_partitioning(0, [])
    return result


# Pythonic solution uses bottom-up construction.
def palindrome_partitioning_pythonic(text):
    return ([[text[:i]] + right for i in range(1, len(text) + 1)
             if text[:i] == text[:i][::-1]
             for right in palindrome_partitioning_pythonic(text[i:])] or [[]])
# @exclude


def check_ans(vecs, input):
    for vec in vecs:
        temp = ''
        for s in vec:
            assert s == s[::-1]
            temp += s
        assert temp == input


def rand_string(length):
    return ''.join(random.choice(string.ascii_lowercase) for _ in range(length))


def simple_test():
    result = palindrome_partitioning('abbbac')
    v0 = ['a', 'b', 'b', 'b', 'a', 'c']
    v1 = ['a', 'b', 'bb', 'a', 'c']
    v2 = ['a', 'bb', 'b', 'a', 'c']
    v3 = ['a', 'bbb', 'a', 'c']
    v4 = ['abbba', 'c']
    golden = [v0, v1, v2, v3, v4]
    assert result == golden
    assert palindrome_partitioning_pythonic('abbbac') == golden


def main():
    simple_test()
    if len(sys.argv) == 2:
        s = sys.argv[1]
        result = palindrome_partitioning(s)
        check_ans(result, s)
        print('string s =', s)
        for vec in result:
            print(*vec)
        result = palindrome_partitioning_pythonic(s)
        check_ans(result, s)
    else:
        for _ in range(1000):
            s = rand_string(random.randint(0, 10))
            result = palindrome_partitioning(s)
            check_ans(result, s)
            print('string s =', s)
            for vec in result:
                print(*vec)
            result = palindrome_partitioning_pythonic(s)
            check_ans(result, s)


if __name__ == '__main__':
    main()
